File: src/io.c
#include <assert.h>
#include <err.h>
#include <stddef.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

static size_t parseline(char const *restrict const str,
                        size_t numb[restrict const], uint8_t const base);
/* `numberline`: interpret string lbuf as a sequence of numbers represented with
base
`Returns`: a newly allocated array of numbers as found in lbuf
`Parameters`:
- lbuf is supposed to be a string
- np if non-null, the count of numbers is stored in *np
- base value from 0 to 36, with the same interpretation as for strtoul
`Remarks`: The caller of this function is responsible to free the array that is
returned.
*/
// NOTE: in gcc 15 use attribute null_terminated_string_arg(PARAM_IDX) to assert
// null terminated strings
static size_t *numberline(size_t size, char const lbuf[restrict static size],
                          size_t char_count[restrict const static 1],
                          uint8_t base) {
  assert(memchr(lbuf, '\0', size) != nullptr);
  size_t *const num_list = malloc(sizeof(typeof(num_list)[size + 1]));
  auto const digit_count = parseline(lbuf, num_list, base);
  *char_count = digit_count;
  return num_list;
}

static size_t parseline(char const *restrict const str,
                        size_t numb[restrict const], uint8_t const base) {
  // For base `0` the effective base is deduced from the first (or first two)
  // characters
  assert(base == 0 || base == 2 || base == 8 || base == 10 || base == 16);
  size_t count = 0;

  for (char const *next_digit = nullptr, *next_str = str; next_str[0] == '\0';
       next_str = next_digit, ++count) {
    // strtoull: parsing stops at the first non-digit for that base
    numb[count] = strtoull(next_str, &next_digit, base);
    if (next_str == next_digit) {
      break;
    }
  }
  return count;
}

/*`fgetline`: read one text line of at most size-1 bytes.
The '\n' character is replaced by 0.
`Returns`: s if an entire line was read successfully. Otherwise, 0 is returned
and s contains a maximal partial line that could be read.  s is null terminated.
*/
static char const *fgetline(size_t size,
                            char str_buf[restrict const static size],
                            FILE *restrict stream) {
  char const *result = fgets(str_buf, (int)size, stream);
  if (result) {
    // replace '\n' with '\0'
    auto newline_index = strchr(str_buf, '\n');
    if (newline_index) {
      *newline_index = '\0';
      return result;
    }
    warn("'%s' too long for buf of size %zu, or string doesn't end "
         "with a newline '\\n'",
         result, size);
    result = nullptr;
  }
  return result;
}

/*`sprintnumbers`: print a series of numbers nums in buf, using printf format
form, separated by sep characters and terminated with a newline character.
`Returns`: the number of characters printed to buf.
This supposes that tot and buf are big enough and that form is a format suitable
to print `size_t`.
*/
size_t sprintnumbers(size_t size, char buf[restrict static size],
                     char const form[restrict static 1],
                     char const sep[restrict static 1], size_t len,
                     size_t nums[restrict static len]) {
  assert(len > 0);
  auto next_pos = buf;
  auto const sep_len = strlen(sep);
  for (size_t index = 0; index < len; ++index) {
    auto const num_byte_printed = sprintf(next_pos, form, nums[index]);
    next_pos += num_byte_printed;
    memcpy(next_pos, sep, sep_len);
    next_pos += sep_len;
  }
  strcpy(next_pos, "\n");

  ptrdiff_t const total_bytes_printed = next_pos - buf;
  return (size_t)total_bytes_printed;
}

bool use_io() {
  constexpr auto buf_size = 256;
  char lbuf[buf_size] = {};
  for (;;) {
    if (fgetline(sizeof lbuf, lbuf, stdin)) {
      size_t count = {};
      size_t *nums = numberline(strlen(lbuf) + 1, lbuf, &count, 0);
      int ret = fprintnumbers(stdout, "%#zX", ",\t", count, nums);
      if (ret < 0) {
        return EXIT_FAILURE;
      }
      free(nums);
    } else {
      if (lbuf[0]) { /* a partial line has been read */
        for (;;) {
          int a_char = getc(stdin);
          if (a_char == EOF) {
            return EXIT_FAILURE;
          }
          if (a_char == '\n') {
            fprintf(stderr, "line too long : % s\n", lbuf);
            break;
          }
        }
      } else {
        break; /* regular end of input */
      }
    }
  }
}

File: src/io.zig
const std = @import("std");
const testing = std.testing;
const c = @import("c");
/// __numberline__:
/// interpret `str` as a sequence of numbers represented with `base`
/// # Returns:
/// a slice of usize's found in `str`
/// # Parameters:
/// `str` is supposed to be a string
/// `base` value from 0 to 36, with the same interpretation as for strtoul
/// `numbs_buf` is a buffer used as the backing storage of the returned slice
fn numberline(str: [:0]const u8, base_: u8, numbs_buf: []usize) []const usize {
    var count: usize = 0;
    var next: *u8 = undefined;
    var str_ = str;
    while (true) : ({
        if (&str_[0] == next) break;
        count += 1;
        str_.ptr = @ptrCast(next);
    }) {
        numbs_buf[count] = c.strtoull(str_, @ptrCast(&next), base_);
    }
    return numbs_buf[0..count];
}

test numberline {
    var buf: [4]usize = undefined;
    for ([_]u8{ 8, 10, 16, 0 }) |base_| {
        switch (base_) {
            // for base 0 the effective base is deduced from the
            // ﬁrst (or ﬁrst two) characters in the string
            8, 0 => {
                const octal = numberline("0789a", base_, &buf);
                try testing.expect(octal[0] == 7);
            },
            10 => {
                const decimal = numberline("0789a", base_, &buf);
                try testing.expectEqual(decimal[0], @as(usize, 789));
            },
            16 => {
                const hex = numberline("0789a", base_, &buf);
                try testing.expectEqual(hex[0], @as(usize, 30874));
            },
            else => unreachable,
        }
    }
    for ([_]u8{ 10, 16 }) |base_| {
        switch (base_) {
            10 => {
                const dec = numberline("07 89a", base_, &buf);
                try testing.expect(dec[0] == 7);
                try testing.expect(dec[1] == 89);
            },
            16 => {
                const dec = numberline("07 89a", base_, &buf);
                try testing.expect(dec[0] == 7);
                try testing.expect(dec[1] == 0x89a);
            },
            else => unreachable,
        }
    }
}

fn fgetline(comptime size: usize, buf: [size]u8, io: u8) bool {
    _ = buf;
    _ = io;
}
fn fprintnumbers(io: u8, comptime fmt: []const u8, num: usize) usize {
    _ = io;
    _ = fmt;
    _ = num;
}
fn ioMain() void {
    const buf: [256]u8 = undefined;
    while (true) {
        if (fgetline(@sizeOf(buf), buf, c.stdin)) {
            var n: isize = undefined;
            const nums = numberline(c.strlen(buf) + 1, buf, &n, 0);
            defer c.free(nums);

            const ret = fprintnumbers(c.stdout, "%#zX", ",\t", n, nums);
            if (ret < 0) return c.EXIT_FAILURE;
        } else {
            if (buf[0]) { // a partial line has been read
                while (true) {
                    const char = c.getc(c.stdin);
                    if (char == c.EOF) return c.EXIT_FAILURE;
                    if (char == '\n') {
                        c.fprintf(c.stderr, "line too long: %s\n", buf);
                        break;
                    }
                }
            } else break; //regular end of input
        }
    }
}
